### ISSUE TITLE
Add support for Nodes system to work with JS code similar to GLSL and WGSL

### ISSUE THREAD

**Issue by LeviPesin** [mrdoob/three.js#25256]
### Description

I suggest to add support for Nodes system to work with JS code similar to GLSL and WGSL. I see multiple ways in which it can be useful:
- It can be useful for doing CPU/GPGPU calculations depending on environment. I.e. you can build a large Nodes program, which then can be executed either by JavaScript, by `WebGPURenderer.compute()`, or `WebGLRenderer.compute()` (I will do a PR on that soon).
- It can be used to replicate the removed `SoftwareRenderer` (a CPU-side renderer).

### Solution

Add support for Nodes system to work with JS code. I can work on this.

### Alternatives

Don't add support for Nodes system to work with JS code.

**Comment by sunag** [comment]
Yeah! And we will be able to use that to create interactive stuff in `NodeEditor`
https://github.com/mrdoob/three.js/pull/23165#issuecomment-1007060716

**Comment by pailhead** [comment]
Hi, I haven't been here in a minute so I'm probably out of the loop... But, why do you think that this should be part of three.js? 

Composition should work with regular js, ie:

```
const myDAG = new MyWhatever(new MyWhatever('foo'))
```

Also you should be able to just unroll the DAG and get regular js/glsl that three **already** understands. In other words, having two node (nocode?) systems may bloat _this_ project, when it could possibly be an (un)related project, in it's own repo. 

If it's of any interest, i played with these topics a bit. I found it fairly trivial to implement an editor (although mine got really complicated :( ) and export three.js materials. It's very hard, maybe even impossible to generate a shader for three though, because it needs to be generated for a specific version range. 
https://dusanbosnjak.com/test/nodes2/

It is also ( i think ) possible to remove all materials from three.js and leave just `ShaderMaterial` or at least it was possible up to 120 or 130 or so. 
https://github.com/pailhead/three-refactor-chunk-material

So, what i'm trying to say, rather than (IMHO) bloating this project, it's actually possible to make it leaner and more specific to a set of problems. In the case of materials it's just validating GLSL. Three did a decent job formatting that GLSL with just string manipulation and regexs, it did not require the whole nodes system.

And then, i think it's possible to make the whole material system much leaner. There is no need for three to know about "Basic" or "Lambert" material, it could know just about "Material". 

My reasoning is that working with nodes requires a nocode editor, otherwise it feels like bloat to me. Just my 2 cents :)

**Comment by LeviPesin** [comment]
The main benefit of the nodes system is flexibility.
You cannot really change an internal function in a GLSL chunk without rewriting it and including as another one. And it will be prone to break on every release.
With nodes system it is just a few lines of JS code (or raw GLSL if you want).
Nodes system doesn't produce code as clean as hand-written indeed... But that can be mostly fixed by inserting `label( ..., name )` at places to make variables variables in GLSL.

> My reasoning is that working with nodes requires a nocode editor, otherwise it feels like bloat to me.

There are such -- apart from Three.js' [base editor](https://threejs.org/editor) there is the [node editor](https://threejs.org/examples/webgpu_nodes_playground.html).

(By the way, I think this isn't very related to this issue)

**Comment by pailhead** [comment]
Right, I was merely asking, why change internals of a chunk at all, or rather, why is there something "internal" here at all :)

If you are proposing more systems (such as nodes) to be part of some future renderer or even `/core` I believe that I'm on the right topic. If i'm totally misunderstanding this, then apologies.

**Comment by mrdoob** [comment]
I'm also confused about where `NodeMaterial` is going. Seems like it's not only about materials anymore.

I've proposed to move the node editor code out of examples: https://github.com/mrdoob/three.js/pull/25692#issuecomment-1477373700
Hopefully that should clarify things a bit.

**Comment by pailhead** [comment]
I think thats fair, but I think the point here is that if you put a jet engine inside of three it could fly, but why, does it need to?

https://github.com/zadvorsky/three.bas

**Comment by sunag** [comment]
> I think this allows you do do gpgpu with just JavaScript, it’s very old though:

I think the concept is a little different from `three.bas`. I think the Node System mission allows you to code TSL and Nodes in any backend or for any need like GPGPU computations. The same code programmed using TSL (Nodes) should work on WebGPU/WebGL and derivatives renderers. NodeMaterial uses the Node System in its base to build the shaders.

> It's very hard, maybe even impossible to generate a shader for three though, because it needs to be generated for a specific version range.

In fact, this is one of the Node System solutions in relation to ShaderChunk.

> I think that in theory you could build any kind of a tool you want in parallel with threejs, and that it could actually yield cleaner code do to separation of concerns.

It is certainly a possibility, I think that over time we add some things and eliminate others and this is part of the innovation and maintenance process. This makes a lot of sense to me as a No-Code Node Graph, @LeviPesin saw another possibility.

**Comment by pailhead** [comment]
> In fact, this is one of the Node System solutions in relation to ShaderChunk.

Just to clarify this - it wouldn’t address this in old versions of three that still rely on chunks? At some point if chunks go away in favor of nodes you are saying they would be compatible with different renderers?

If so, I’d argue that the contract is still just a valid, or somewhat valid GLSL string. My reasoning is that there could be several different node standards or projects that could output this string.

**Comment by sunag** [comment]
@pailhead Comparing strings is a rationale for ShaderChunk not Node System, WebGPURenderer does not have ShaderChunk, there are no strings to compare there, there are no hacks, so it does not suffer from these problems.

**Comment by pailhead** [comment]
Just found an issue of mine that may be related. If i'm understanding this correctly, it's proposing the complete opposite direction of what's being proposed here.

https://github.com/mrdoob/three.js/issues/15537

**Comment by pailhead** [comment]
@sunag whoa, yeah I haven't worked with the WebGPURenderer at all, you're saying it doesn't work with `ShaderMaterial` at all?

**Comment by sunag** [comment]
> I don't actually know what "chunks" refer to, i think of that collection of strings that can be assembled into shaders, but i see a lot of variables named chunks in this code. I was not referring to those big "chunks" just these things that the renderer appends or switches out.

This approch is discontinued in WebGPURenderer, Node System does not use any type .replace() or indexOf() in its finalization, that's why we have Node API for Lights for example. Normally these string manipulations are hacks, that's because we forced the creation of a resource that wasn't contemplated at the API level, that's why we have so many side effects.

**Comment by LeviPesin** [comment]
> I don't actually know what "chunks" refer to, i think of that collection of strings that can be assembled into shaders, but i see a lot of variables named chunks in this code. I was not referring to those big "chunks" just these things that the renderer appends or switches out. My assumption was that this code would remain in some form, i may be totally wrong about this.

This is what I meant by "You cannot really change an internal function in a GLSL chunk without rewriting it and including as another one. And it will be prone to break on every release".
GLSL chunks that WebGLRenderer uses are very inter-dependent and have large possibility to break if one is missed or used in incorrect order. Moreover, users struggle with hacks like `onBeforeCompile( replaceShaders )` to modify these chunks, increasing a possibility of breakage.
Nodes system doesn't have this problem.

I suggest we move this discussion to https://github.com/mrdoob/three.js/issues/16440. I think that issue also partially answers to your question why nodes system is located in Three.js repository -- because we want to have it in the core (ideally replacing shader chunks API 🙏) at some point.

**Comment by LeviPesin** [comment]
And I also think it makes more sense now to consider node system not as tool for working with and customizing materials, but as a tool for creating language-independent backend-independent code with high flexibility. This allows making node materials, GPGPU computations, et cetera...

**Comment by pailhead** [comment]
Ok fair enough, that’s a pretty big change compared to the current architecture. Thank you for educating me. 

I agree that in this context this makes sense. Threes JavaScript apis also tends to break between releases and a node based system to replace it and make it version independent would be great.
